docs/docs/reference/metaprogramming/macros.md

```diff
@@ -614,4 +614,81 @@ compilation of the suspended files using the output of the previous (partial) co
 In case all files are suspended due to cyclic dependencies the compilation will fail with an error.
 
 
+### Pattern matching on quoted expressions
+
+It is possible to deconstruct or extract values out of `Expr` using pattern matching.
+
+#### scala.quoted .matching
+
+In `scala.quoted.matching` contains object that can help extract values from `Expr`.
+
+* `scala.quoted.matching.Const`: matches an expression a literal value and returns the value.
+* `scala.quoted.matching.ExprSeq`: matches an explicit sequence of expresions and returns them. These sequences are useful to get individual `Expr[T]` out of a varargs expression of type `Expr[Seq[T]]`.
+* `scala.quoted.matching.ConstSeq`:  matches an explicit sequence of literal values and returns them.
+
+These could be used in the following way to optimize any call to `sum` that has statically known values.
+```scala
+inline def sum(args: =>Int*): Int = ${ sumExpr('args) }
+private def sumExpr(argsExpr: Expr[Seq[Int]])(given QuoteContext): Expr[Int] = argsExpr.underlyingArgument match {
+  case ConstSeq(args) => // args is of type Seq[Int]
+    Expr(args.sum) // precompute result of sum
+  case ExprSeq(argExprs) => // argExprs is of type Seq[Expr[Int]]
+    val staticSum: Int = argExprs.map {
+      case Const(arg) => arg
+      case _ => 0
+    }.sum
+    val dynamicSum: Seq[Expr[Int]] = argExprs.filter {
+      case Const(_) => false
+      case arg => true
+    }
+    dynamicSum.foldLeft(Expr(staticSum))((acc, arg) => '{ $acc + $arg })
+  case _ =>
+    '{ $argsExpr.sum }
+}
+```
+
+#### Quoted patterns
+
+Quoted pattens allow to deconstruct complex code that contains a precise structure, types or methods.
+Patterns `'{ ... }` can be placed in any location where Scala expects a pattern.
+
+For example
+```scala
+optimize {
+  sum(sum(1, a, 2), 3, b)
+} // should be optimized to 6 + a + b
+```
+
+```scala
+def sum(args: =>Int*): Int = args.sum
+def optimize(arg: Int): Int = ${ optimizeExpr('arg) }
+def optimizeExpr(body: Expr[Int])(given QuoteContext): Expr[Int] = body match {
+  // Match a call to sum without any arguments
+  case '{ sum() } => Expr(0)
+  // Match a call to sum with an argument $n of type Int. n will be the Expr[Int] representing the argument.
+  case '{ sum($n) } => n
+  // Match a call to sum and extracts all its args in an `Expr[Seq[Int]]`
+  case '{ sum($args: _*) } => sumExpr(args)
+  case body => body
+}
+def sumExpr(args: Expr[Seq[Int]])(given QuoteContext): Expr[Int] = args.underlyingArgument match {
+    def flatSumArgs(arg: Expr[Int]): Seq[Expr[Int]] = arg match {
+      case '{ sum($subArgs: _*) } => subArgs.flatMap(listflatSumArgsSum)
```

Comment:
I think this code sample may be incomplete. I don't see `listFlatSumArgsSum` anywhere other than here, and the same seems really specific for part of the quoting framework.

Reply:
That is probably `flatSumArgs`. I will add a test case to make sure the code runs as intended and update the example.